posts/2020-11-27-typing-your-actors.adoc

```diff
@@ -0,0 +1,389 @@
+            Behaviors.same
+          case WSFailure(t) =>
+            context.log.error("WSFailure: {}", t)
+            Behaviors.same
+        }
+      }
+      .receiveSignal {
+        case (_, Terminated(ref)) =>
+          val newData = RoomManagerData(data.rooms.filter {
+            case (_, roomRef) => roomRef != ref
+          })
+          receiveBehaviour(newData)
+      }
+----
+
+Room:
+
+[source,scala]
+----
+final case class RoomData(users: List[User], currentIssue: String, issueLastEditBy: Option[UUID])
+
+  object RoomData {
+    val empty: RoomData = RoomData(List.empty[User], "", Option.empty[UUID])
+  }
+
+  def apply(roomId: UUID): Behavior[Command] =
+    Behaviors.setup[Command] { _ =>
+      receiveBehaviour(roomId, RoomData.empty)
+    }
+
+  private def receiveBehaviour(roomId: UUID, data: RoomData): Behavior[Command] =
+    Behaviors.receive[Command] { (context, message) =>
+      message match {
+        case Join(user) =>
+          val newData = data.copy(users = user :: data.users)
+          setupNewUser(user, roomId, newData)
+          broadcast(WSMessage(MessageType.Join, roomId, user.id, user.name), newData.users, context)
+          receiveBehaviour(roomId, newData)
+        case Vote(userId, estimation) =>
+          val newUsers = data.users.map { u =>
+            if (userId == u.id) u.copy(voted = true, estimation = estimation)
+            else u
+          }
+          broadcast(WSMessage(MessageType.Vote, roomId, userId, estimation), newUsers, context)
+          receiveBehaviour(roomId, data.copy(users = newUsers))
+        case ClearVotes(userId) =>
+          val newUsers = data.users.map(_.copy(voted = false, estimation = ""))
+          broadcast(
+            WSMessage(MessageType.Clear, roomId, userId, WSMessage.NoExtra),
+            newUsers,
+            context
+          )
+          receiveBehaviour(roomId, data.copy(users = newUsers))
+        case ShowVotes(userId) =>
+          broadcast(
+            WSMessage(MessageType.Show, roomId, userId, WSMessage.NoExtra),
+            data.users,
+            context
+          )
+          Behaviors.same
+        case Leave(userId) =>
+          val newUsers = data.users.filter(_.id != userId)
+          broadcast(
+            WSMessage(MessageType.Leave, roomId, userId, WSMessage.NoExtra),
+            newUsers,
+            context
+          )
+          if (newUsers.isEmpty) {
+            Behaviors.stopped
+          } else {
+            Behaviors.same
+          }
+        case EditIssue(userId, issue) =>
+          broadcast(WSMessage(MessageType.EditIssue, roomId, userId, issue), data.users, context)
+          receiveBehaviour(
+            roomId,
+            data.copy(currentIssue = issue, issueLastEditBy = Option(userId))
+          )
+      }
+
+    }
+----
+
+Takeaways:
+
+* Internal state is now immutable.
+* Sender reference is not present, causing modification on CreateRoom.
```

Comment:
In addition to what @ckipp01 suggests, I would be more explicit in saying that as `sender()` is no longer available, the `ActorRef` to the sender of the message, is passed as part of the incoming message.